### ISSUE TITLE
per-package setting of compiler flags

### ISSUE THREAD

**Issue by davydden** [spack/spack#3145]
I know that one can set compiler flags either in `compilers.yaml` or on the command line `spack install foo cppflags="-march=native"`. Additionally, I saw some packages add/specify developer-recommended flags during configure stage (don't recall which, though). 

I am trying to figure out what is the best way to use non-default flags for a single package only (manually i would do that with `DCMAKE_CXX_FLAGS="-march=native"` during `cmake` config). I could achieve that with
```
spack install --only dependents foo
spack install foo cppflags="-march=native"
```
But that's not ideal. It is also not easy to extend to the case when custom flags have to be added to some dependencies of the DAG.

Keeping this in mind, I think it would be useful to be able to specify extra per-package flags in `packages.yaml`:
```
packages:
  foo:
    cppflags: 
       gcc: -march=native
       clang: -march=native -ffp-contract=fast
```
Any thoughts?

p.s. One can of course always hard-code extra developer's recommended flags in `package.py`.

**Comment by citibeth** [comment]
I believe #3142 does what you are asking for?

> Any thoughts?
> p.s. One can of course always hard-code extra developer's recommended flags in `package.py`.

I believe this is the best.  Unless these extra flags are only useful for you?  It should be possible to subclass packages... so you can create in your own repo, `davydden-package-x`, which is a subclass of the regular `package-x` that then sets additional compiler flags, etc.

**Comment by davydden** [comment]
> I believe #3142 does what you are asking for?

one can kind of make it work with a separate config setting, but it's not any better than adding compiler flags in the command line. So, i would say the PR you cite does not solve this problem.

> Unless these extra flags are only useful for you?

no, they are not only for myself.

**Comment by citibeth** [comment]
@adamjstewart So it sounds like what we need is a way to add compiler flags
from `package.py`.  Does such a mechanism exist?  If not, how hard would it
be to add?

**Comment by adamjstewart** [comment]
> So it sounds like what we need is a way to add compiler flags from `package.py`. Does such a mechanism exist? If not, how hard would it be to add?

You can set `CPPFLAGS` and friends from a `package.py`. We do it all the time. As far as I know, there is no way to append to the flags that get sent directly to the compiler wrapper though.

**Comment by adamjstewart** [comment]
I wonder if you can append to `self.compiler.cppflags`. I've never seen it done before though.

**Comment by davydden** [comment]
> So it sounds like what we need is a way to add compiler flags from `package.py`.

that would be irrelevant for my usage because the package uses `CMake` and I can always append extra flags by configuring with 
```
-DCMAKE_CXX_FLAGS="-march=native"
```

**Comment by alalazo** [comment]
@becker33 I think this has been fixed by #6415, can you confirm it?

**Comment by bebosudo** [comment]
Hi! I have the same feature request; I'm using spack v0.11.2.

#3142 isn't merged, and I don't understand how #6415 works.
I tried adding to my `packages.yaml`:
```
packages:
  netcdf:
    flags:
      cppflags: "-fPIC"
```

but it crashed with a python error:
```
==> Error: 'KeysView' object has no attribute 'index'
```

Fortunately, I solved my specific problem by using the variant `+pic`:
```
packages:
  netcdf:
    variants: +pic
```

but changing flags on a package basis would be great.

**Comment by scheibelp** [comment]
> I don't understand how #6415 works.
> I'm using spack v0.11.2.

The functionality in #6415 is not available from `0.11.2`, would you be able to try the latest develop?

#6415 is documented in the "Compiler Flags" section of the packaging guide: https://spack.readthedocs.io/en/latest/packaging_guide.html#compiler-flags. To make use of that, you have to edit the package.py file for a package, vs. editing `packages.yaml`.

If you are building a package that always needs a particular flag, regardless of what system you build it on, then overriding the `flag_handler` method is the way to go. It is also important if you want to set any flags (including in `compilers.yaml`) and the build system needs to know about compiler flags (e.g. `-DCMAKE_C_FLAGS` for CMake).

`pic` is generally an option that users express in the package.py file (and it is controlled with a variant).

**Comment by bebosudo** [comment]
> The functionality in #6415 is not available from 0.11.2, would you be able to try the latest develop?

Sure, I can switch to the development version.

> To make use of that, you have to edit the package.py file for a package, vs. editing packages.yaml.

Okay, thanks for pointing me to the documentation.. A question: is this planned to stay as it is also in the future?

Regarding the pic option, I submitted #9605 to fix the netcdf packages.

**Comment by alalazo** [comment]
Solved in #6415
